File: server/pet/combat_inputs.go
package pet

import "errors"

var ErrInvalidCombatInputs = errors.New("invalid pet combat inputs")

// CombatInputs is the complete Pet Care -> combat C5 boundary. Combat owns
// every formula that consumes these integers; Pet Care only supplies the
// replay-owned Trust value beside the Founder-owned Soul value.
type CombatInputs struct {
	PetTrustPPM int64 `json:"pet_trust_ppm"`
	Soul        int64 `json:"soul"`
}

func NewCombatInputs(petTrustPPM, soul int64) (CombatInputs, error) {
	if petTrustPPM < 0 || petTrustPPM > 1_000_000 || soul < -maxExactInteger || soul > maxExactInteger {
		return CombatInputs{}, ErrInvalidCombatInputs
	}
	return CombatInputs{PetTrustPPM: petTrustPPM, Soul: soul}, nil
}

func CombatInputsForState(state CareState, soul int64) (CombatInputs, error) {
	return NewCombatInputs(state.TrustPPM, soul)
}

File: client/src/pet/combat-inputs.ts
import { MAX_EXACT_INTEGER } from "../numeric";
import type { PetCareState } from "./state";

export interface PetCombatInputs {
  readonly pet_trust_ppm: number;
  readonly soul: number;
}

export function petCombatInputs(petTrustPpm: number, soul: number): PetCombatInputs {
  if (!Number.isSafeInteger(petTrustPpm) || petTrustPpm < 0 || petTrustPpm > 1_000_000 ||
    !Number.isSafeInteger(soul) || soul < -MAX_EXACT_INTEGER || soul > MAX_EXACT_INTEGER) {
    throw new RangeError("invalid pet combat inputs");
  }
  return { pet_trust_ppm: petTrustPpm, soul };
}

export function petCombatInputsForState(state: PetCareState, soul: number): PetCombatInputs {
  return petCombatInputs(state.trust_ppm, soul);
}
